### ISSUE TITLE
Empty OSNotification Handler?!?!

### ISSUE THREAD

**Issue by HackShitUp** [OneSignal/OneSignal-iOS-SDK#161]
Below I have the follow code:

```Swift
OneSignal.initWithLaunchOptions(launchOptions,
                                        appId: "571bbb3a-3612-4496-b3b4-12623256782a",
                                        handleNotificationReceived: { (notification) in
                                            
                                            // This block gets called when the user reacts to a notification received
                                            let payload = notification?.payload
                                            let fullMessage = payload?.title
                                            
                                            print("\(fullMessage)")
                                            
                                            // Banner for notification
                                            if fullMessage!.hasPrefix("\(PFUser.current()!.username!.uppercased())") {
                                                // If PFUser.currentUser()! caused
                                                // Sent notification
                                                // Set "invisible banner"
                                                let banner = Banner(title: "",
                                                                    subtitle: "",
                                                                    image: nil,
                                                                    backgroundColor: UIColor.clear
                                                )
                                                
                                                banner.dismissesOnTap = true
                                                banner.show(duration: 0.0)
                                                
                                                
                                            } else if chatUsername.count != 0 && chatUserObject.count != 0 {
                                                if fullMessage!.hasPrefix("from") && fullMessage!.hasSuffix("\(chatUsername.last!.uppercased())") {
                                                    
                                                    // if notificaiton titles: "from <Username>"
                                                    // and PFUser.currentUser! is CURRENTLY talking to OtherUser...
                                                    // Reload data for SlackChat
                                                    NotificationCenter.default.post(name: rpChat, object: nil)
                                                }
                                                
                                            } else {
                                                // Set visible banner
                                                let banner = Banner(title: "\(fullMessage!)",
                                                    subtitle: "",
                                                    image: UIImage(named: "RedplanetLogo"),
                                                    backgroundColor: UIColor(red:1.00, green:0.00, blue:0.31, alpha:1.0)
                                                )
                                                banner.dismissesOnTap = true
                                                banner.show(duration: 3.0)
                                            }
                                            
                                            
            }, handleNotificationAction: { (result) in
               // Action
                
            }, settings: [kOSSettingsKeyAutoPrompt : false, kOSSettingsKeyInFocusDisplayOption : false])
```

It handles the code when a notification was received in AppDelegate. The postNotification function is fired fine, the receiving of the notification however, results in nil. I've checked my OneSignal dashboard and it's showing me that the API Notifications were sent successfully. All the following protocols actually result in nil:

• notification?.payload
• notification?.payload?.title

I handle the received notification functions based on its title, but it returns the following error:

`fatal error: unexpectedly found nil while unwrapping an Optional value`

**Comment by jkasten2** [comment]
@HackShitUp `notification?.payload` should always be set. Can you try a basic notification with just a message body and title? Can you let us know which version of our SDK you are using and the iOS version you are seeing the issue on?

Thanks.

**Comment by HackShitUp** [comment]
@jkasten2 Hey, my name is Josh too by the way haha.

This is a basic notification here's the code for posting the notification:
`                                                                // MARK: - OneSignal
                                                                // Send push notification
                                                                OneSignal.postNotification(
                                                                    ["contents":
                                                                        ["en": "\(PFUser.current()!.username!.uppercased()) shared your Text Post"],
                                                                     "include_player_ids": ["\(self.userObject!.value(forKey: "apnsId") as! String)"]
                                                                    ]
                                                                )`

Again, this fires 100% without error. It logs and even the OS dashboard is telling me it was sent successfully. Here's some more information

- OneSignal SDK Version 2.3.4
- Device iOS: 10.1.0

**Comment by HackShitUp** [comment]
Also, I'm using Cocoapods

**Comment by HackShitUp** [comment]
I just tested again, and

`notification?.payload.title`

is consistently nil

**Comment by jkasten2** [comment]
@HackShitUp The `title` from `notification.payload` will be `nil` from looking at the REST API call you made. You will need to set it on your notification as a `headings` field.
```swift
OneSignal.postNotification(["contents": ["en": "\(PFUser.current()!.username!.uppercased()) shared your Text Post"], "headings": ["en": "Your title here!"], "include_player_ids": ["\(self.userObject!.value(forKey: "apnsId") as! String)"] ] )
```

**Comment by HackShitUp** [comment]
(1) Is there documentation for this? I'm aware this is in JSON but I'm not too familiar with this.
(2) Because I'd like to have the notification displayed in a thinner font. The code you suggest prevents the crash and it works. However, the notification when the app is not in the foreground shouldn't be the title (bold). How do I make sure it's in the body (thinner text font for iOS)?

**Comment by HackShitUp** [comment]
@jkasten2

**Comment by HackShitUp** [comment]
@jkasten2 The effective solution was to set the following protocols when handling the OSNotification in AppDelegate:

`notification?.payload.body`

Sorry for the inconvenience, I feel like an idiot. It worked a few days ago but it failed to work after updating the framework. Why is that?

**Comment by jkasten2** [comment]
@HackShitUp Sorry about the confusion, there was a bug where title would set to body text in some cases. This was corrected in our [2.2.1 SDK](https://github.com/OneSignal/OneSignal-iOS-SDK/releases/tag/2.2.1).

Thanks.
